src/graph/backend/elyzor/compiler_loader.cpp

```diff
@@ -0,0 +1,134 @@
+/*******************************************************************************
+ * Copyright 2021-2024 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+#include "compiler_loader.hpp"
+#include "elyzor_backend.hpp"
+
+#ifdef _WIN32
+#include <windows.h>
+#define DNNL_GC_LIB_NAME "graph_compiler.dll"
+#define OPEN_LIB(libname, flags) LoadLibrary(libname)
+#define LOAD_FUNC(handle, funcname) GetProcAddress((HMODULE)handle, funcname)
+#define FREE_LIB(handle) FreeLibrary((HMODULE)handle)
+#define OPEN_LIB_REPORT_ERROR() ("could not find " #DNNL_GC_LIB_NAME)
+#else
+#define OPEN_LIB(libname, flags) dlopen(libname, flags)
+#define LOAD_FUNC(handle, funcname) dlsym(handle, funcname)
+#define FREE_LIB(handle) dlclose(handle)
+#define OPEN_LIB_REPORT_ERROR() dlerror()
+#endif
+#ifdef __APPLE__
+#include <dlfcn.h>
+#define DNNL_GC_LIB_NAME "libgraph_compiler.dylib"
+#elif !defined(_WIN32)
+#include <dlfcn.h>
+#define DNNL_GC_LIB_NAME "libgraph_compiler.so"
+#endif
```

Comment:
Why not using macros for windows only? It's not only more compact:
<details>

```c++
#ifdef _WIN32
#include <windows.h>
#define DNNL_GC_LIB_NAME "graph_compiler.dll"
#define dlopen(libname, flags) LoadLibrary(libname)
#define dlsym(handle, funcname) GetProcAddress((HMODULE)handle, funcname)
#define dlclose(handle) FreeLibrary((HMODULE)handle)
#define dlerror() "could not find " #DNNL_GC_LIB_NAME
#elif __APPLE__
#include <dlfcn.h>
#define DNNL_GC_LIB_NAME "libgraph_compiler.dylib"
#else
#include <dlfcn.h>
#define DNNL_GC_LIB_NAME "libgraph_compiler.so"
#endif
```

</details>

It makes code much more readable:
```C++
handle_ = OPEN_LIB(libname_, RTLD_LAZY | RTLD_DEEPBIND);
```
vs
```C++
handle_ = dlopen(libname_, RTLD_LAZY | RTLD_DEEPBIND);
```

Reply:
> Why not using macros for windows only?

So it would be visible that it's a macro and not an actual call. If I see `dlopen(...)` in the code I would automatically assume that it's an actual `dlopen` call (which may be not in case of windows)

Reply:
When I see `dlopen(...)`, I understand what it does. In case of the macros, I need to find the macros definition first.

Reply:
Anyway, it's a matter of taste. Would be good to see other opinions.

Reply:
Well, I was aiming to match the level-zero's logic of handling dynamically loaded libs and they're [using custom macroses](https://github.com/oneapi-src/level-zero/blob/d8256c4d5f4733a7b31115412ca1022bb2640585/source/inc/ze_util.h#L17-L52) as well (instead of redefining dlopen func)

Reply:
As I said, it's a matter of taste.
https://github.com/oneapi-src/oneTBB/blob/762b8af7e2d15af8246a88815d578eb1ef1ab548/src/tbb/dynamic_link.cpp#L35

https://github.com/search?q=%22%23define+dlopen%28%22&type=code

Reply:
don't have a strong preference here, let's use macroses only for windows